File: 2021/src/day_8.rs
use std::fs;

pub const TITLE: &str = "--- Day 8: Seven Segment Search ---";

pub fn part_1() -> i64 {
    fs::read_to_string("inputs/day8")
        .expect("Unable to read input")
        .lines()
        .map(|line| line.split(" | ").skip(1).next().unwrap())
        .map(|output| output.split(" "))
        .flatten()
        .filter(|&each| match each.len() {
            2 | 4 | 3 | 7 => true,
            _ => false,
        })
        .count() as i64
}

fn unique_chars(s1: &str, s2: &str) -> String {
    s1.chars()
        .filter(|&each| !s2.contains(each))
        .collect::<String>()
}

fn convert(input: Vec<&str>, output: Vec<&str>) -> i64 {
    // Sort each string in the vector
    let sorted_output: Vec<String> = output
        .iter()
        .map(|&each| {
            let mut chars = each.chars().collect::<Vec<char>>();
            chars.sort_by(|a, b| a.cmp(b));

            String::from_iter(chars)
        })
        .collect();

    let mut input_sorted_by_length: Vec<&str> = input.clone();
    input_sorted_by_length.sort_unstable_by(|a, b| a.len().cmp(&b.len()));

    let mut mappings = vec![""; 10];
    mappings[1] = input_sorted_by_length[0];
    mappings[4] = input_sorted_by_length[2];
    mappings[7] = input_sorted_by_length[1];
    mappings[8] = input_sorted_by_length[9];

    let mut top_right = "".to_string();
    let mut bottom_right = "".to_string();

    for i in 6..9 {
        if unique_chars(input_sorted_by_length[i], input_sorted_by_length[0].clone()).len() == 5 {
            mappings[6] = input_sorted_by_length[i];
            top_right = unique_chars(input_sorted_by_length[9], mappings[6].clone())
                .chars()
                .next()
                .unwrap()
                .to_string();
            bottom_right = unique_chars(mappings[1], top_right.as_str())
                .chars()
                .next()
                .unwrap()
                .to_string();
            break;
        }
    }

    for i in 3..6 {
        if input_sorted_by_length[i].contains(&top_right)
            && input_sorted_by_length[i].contains(&bottom_right)
        {
            mappings[3] = input_sorted_by_length[i];
        } else if input_sorted_by_length[i].contains(&top_right) {
            mappings[2] = input_sorted_by_length[i];
        } else {
            mappings[5] = input_sorted_by_length[i];
        }
    }

    for i in 6..9 {
        if input_sorted_by_length[i] == mappings[6] {
            continue;
        } else {
            if unique_chars(input_sorted_by_length[i], mappings[4].clone()).len() == 2 {
                mappings[9] = input_sorted_by_length[i];
            } else {
                mappings[0] = input_sorted_by_length[i];
            }
        }
    }

    let mut actual_output = "".to_string();
    let sorted_mappings: Vec<String> = mappings
        .iter()
        .map(|&each| {
            let mut chars = each.chars().collect::<Vec<char>>();
            chars.sort_by(|a, b| a.cmp(b));

            String::from_iter(chars)
        })
        .collect();

    for digit in sorted_output {
        actual_output.push_str(
            sorted_mappings
                .iter()
                .position(|each| each.as_str() == digit)
                .unwrap()
                .to_string()
                .as_str(),
        );
    }

    actual_output.parse::<i64>().unwrap()
}

pub fn part_2() -> i64 {
    fs::read_to_string("inputs/day8")
        .expect("Unable to read input")
        .lines()
        .map(|line| {
            let mut pair = line.split(" | ");
            let input = pair.next().unwrap().split(" ").collect::<Vec<&str>>();
            let output = pair.next().unwrap().split(" ").collect::<Vec<&str>>();
            convert(input, output)
        })
        .sum()
}

File: 2021/python/day_8.py
def part_1():
    counter = 0
    with open("inputs/day8", "r") as file:
        for line in file.read().splitlines():
            counter += len(
                list(
                    filter(
                        lambda x: len(x) in [2, 4, 3, 7],
                        line.split(" | ")[1].split(" "),
                    )
                )
            )
        return counter


def convert(input, output) -> int:
    output = list(map("".join, map(sorted, output)))

    # Map the unique values we know to be 1, 4, 7 and 8 by length of input
    input.sort(key=lambda x: len(x))
    one, four, seven, eight = input[0], input[2], input[1], input[9]
    zero, two, three, five, six, nine = "", "", "", "", "", ""

    top_right, bottom_right = "", ""
    for i in range(6, 9):
        # We get six if after removing all symbols of "1" gives us 5 remaning symbols
        if len(set(input[i]) - set(input[0])) == 5:
            six = input[i]
            top_right = (set(input[9]) - set(input[i])).pop()
            bottom_right = (set(one) - set(top_right)).pop()
            break

    for i in range(3, 6):
        if top_right in input[i] and bottom_right in input[i]:
            three = input[i]
        elif top_right in input[i]:
            two = input[i]
        else:
            five = input[i]

    for i in range(6, 9):
        if input[i] == six:
            continue
        else:
            if len(set(input[i]) - set(four)) == 2:
                nine = input[i]
            else:
                zero = input[i]

    mappings = list(
        map(
            "".join,
            map(sorted, [zero, one, two, three, four, five, six, seven, eight, nine]),
        ),
    )

    display = ""
    for digit in output:
        display += str(mappings.index(digit))

    return int(display)


def part_2():
    total = 0
    with open("inputs/day8", "r") as file:
        for line in file.read().splitlines():
            pair = line.split(" | ")
            total += convert(pair[0].split(" "), pair[1].split(" "))

        return total


print(part_1())
print(part_2())
